src/Compilers/CSharp/Portable/BoundTree/Statement.cs

```diff
@@ -310,20 +320,49 @@ partial class BoundStateMachineScope
         protected override OperationKind StatementKind => OperationKind.None;
     }
 
-    partial class BoundLocalDeclaration : IVariableDeclarationStatement, IVariable
+    partial class BoundLocalDeclaration : IVariableDeclarationStatement
     {
-        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables => ImmutableArray.Create<IVariable>(this);
-
-        ILocalSymbol IVariable.Variable => this.LocalSymbol;
+        private static readonly ConditionalWeakTable<BoundLocalDeclaration, IVariable> s_variablesMappings =
+            new ConditionalWeakTable<BoundLocalDeclaration, IVariable>();
 
-        IExpression IVariable.InitialValue => this.InitializerOpt;
+        ImmutableArray<IVariable> IVariableDeclarationStatement.Variables
+        {
+            get
+            {
+                var builder = ArrayBuilder<IVariable>.GetInstance(1);
+                var variable = s_variablesMappings.GetValue(this, decl => new VariableDeclaration(decl.LocalSymbol, decl.InitializerOpt, decl.Syntax));
+                builder.Add(variable);
+                return builder.ToImmutableAndFree();
```

Comment:
Sure. But since `ImmutableArray` is a value type, I think they are considered same as long as the elements are same.